man/portablectl.xml

```diff
@@ -305,7 +305,13 @@
         profile drop-ins are symlinked while unit files are copied. Note that this option expresses a preference only,
         in cases where symbolic links cannot be created — for example when the image operated on is a raw disk image,
         and hence not directly referentiable from the host file system — copying of files is used
-        unconditionally.</para></listitem>
+        unconditionally.</para>
+
+        <para>When unit files and drop-ins are copied, if the kernel supports it, fsverity will be used to enable integrity
+        protection on those files. If a pkcs7 signature of the fsverity digest is available as <filename>foo.service.p7s</filename>
```

Comment:
hm, uh, oh. not sure how i like this.

sure, with this the contents of the service files will be authenticated, but not the names, hence what's the benefit? i could copy any service file from the image (e.g. reboot.service) and then name it anything else (e.g. httpd.service), and everything will be fucked.

Also the ".p7s" suffix is a bit vague, it should be ".fsverity.p7s" i guess, since it's not clear at all that this is about verity signatures otherwise. People might otherwise assume it was just a regular pkcs7 sig of the contents, which it clearly is not here though.

Also, why not make this an xattr? it's usually how security attributes of files are associated with the files, and that would make a ton of sense for this too, no?

also, it would simplify handling a log, since we could just pass around a single fd, and use it.

Reply:
it might make sense to add a new setting to [Unit] section in unit files: "Name="

it would have to carry the name of the unit file itself. that way you could attack the renaming issue: if people use this fsverity stuff they'd also have to say inside the units what this is about, and thus service files cannot be reappropriated anymore

Reply:
> sure, with this the contents of the service files will be authenticated, but not the names, hence what's the benefit? i could copy any service file from the image (e.g. reboot.service) and then name it anything else (e.g. httpd.service), and everything will be fucked.

The point is what _we_ do, not what a random user with root could do. portabled is authorized by policy to create those files. It's up to the system builder/admin to ensure the security policy doesn't allow writing stuff to /etc, if that's not desired.

> Also the ".p7s" suffix is a bit vague, it should be ".fsverity.p7s" i guess, since it's not clear at all that this is about verity signatures otherwise. People might otherwise assume it was just a regular pkcs7 sig of the contents, which it clearly is not here though.

The portable image format is specific and documented, so I don't see why there would be any such assumption, as there's no other use. Also the normal convention is to append p7s to the file it applies to, without extra suffixes. Most importantly, I'm already using these signatures and using this naming convention, so I cannot change it.

> Also, why not make this an xattr? it's usually how security attributes of files are associated with the files, and that would make a ton of sense for this too, no?

It could be another mechanism sure, but I do need the separate files as well. xattrs are a massive pain when building, especially on windows.

Reply:
.p7s is not a suffix we invented. we don't own it. we cannot just say "it's signatures for fsverity files only". See here for example: https://extension.nirsoft.net/p7s

And we use the .roothash.p7s and .usrhash.p7s suffix for the dm-verity case btw, not just ".p7s".

Reply:
yes, because the file is foo.roothash, so it gets foo.roothash.p7s, as the convention is that the signature is the name of the file it accompanies plus the suffix

Reply:
exactly, "x.p7s" is the signature for the binary data in "x". Which is not the case here. "foo.service.p7s" does not contain the signature for the binary data in "foo.service" Instead, it contains the signature for the fsverity for the binary data in "foo.service". 

i.e. for roothash.p7s i can use openssl from the cmdline to generate and verify the sigs. i cannot do this for foo.service.p7s, i have to use the fsverity tools to get the right binary data to sign. hence it should have a separate suffix.

Reply:
The semantics are arbitrary, having to use openssl rather than fsverity or the other way around is entirely arbitrary and up to whoever defines the usage. The suffix is used to denote which files it goes together with, not the tool that is used with it.

Reply:
> sure, with this the contents of the service files will be authenticated, but not the names, hence what's the benefit?

Note that this pull request does not actually authenticate the file contents either, as it is using the fsverity "builtin signatures" (which is not recommended) and doing so incorrectly --- see my other comment.

Reply:
It does. What you meant to say is that it does so in a way that google doesn't like - unfortunate, but it is not a requirement to have google's approval.

Reply:
It does not, unless the sysctl `fs.verity.require_signatures=1` is turned on (never mentioned in the pull request) *and* something enforces that specific files have fsverity enabled (also not mentioned in the pull request, and I think it would *have* to be included since `systemd` would need to do this?).

"Google" indeed doesn't like the fsverity builtin signatures anymore --- that's because "Google" tried them and they have some major issues.  It's fundamentally not a good approach.  Similarly, they were tried for https://fedoraproject.org/wiki/Changes/FsVerityRPM which ended up being rejected and abandoned.  Maybe you should learn from other people's experiences instead of saying they don't matter.

Of course, IPE would introduce a new use of the builtin signatures which could make them useful in this context.  I personally don't like how IPE duplicates IMA, but disregarding that, at least it would be a design that achieves something.  I'm not sure why you are trying to claim that this pull request has nothing to do with IPE.  (Maybe because you know the systemd maintainers would not accept code that is only useful in combination with non-upstream kernel patches...)

Reply:
It's not a good approach _for google's use case_. You keep making this assumption that google's use cases are the be-all and end-all of all use cases, and nobody could possibly need anything different than what almighty google needs. I am afraid that is not quite the case. I understand that google has an agenda here and the intention is to deprecate this UAPI (https://lore.kernel.org/linux-integrity/20221208033548.122704-1-ebiggers@kernel.org/), but as I said on the LKML, that's not going to work. It is very much needed and in use as-is, and the kernel's userspace interface compatibility cannot simply be broken because google feels like it. Disparaging other projects simply because they don't fit within google's narrow user cases won't help.

Reply:
This has nothing to do with Google or anyone else's employer here, but rather this pull request not actually describing a logical use case.  The reason I'm commenting on this pull request is that I maintain fsverity, and it's important that a project as big as `systemd` adds support for fsverity in some way, it needs to actually make sense, otherwise it will cause all sorts of problems.  I'm already having problems with `systemd` and fscrypt because they are not using the fscrypt API properly.

BTW, I don't know if you noticed, but I listened to your feedback and did not deprecate the builtin signatures.

Reply:
And to be clear, it's not just "no use case described", but actually "it's impossible for this systemd feature to actually be used currently".  If I'm wrong then please correct me --- not just in GitHub comments, which don't really matter, but in the actual code and documentation...

Also, if you want to give IPE as the use case (which I think it is if you were being honest), I can accept that even though I'm not super enthusiastic about IPE.  Of course, it's the systemd maintainers' call whether they want to accept a feature that is only useful with out-of-tree kernel patches.

Reply:
Once more, it can be used just fine, the test is there to prove it

Reply:
No, the test does not show that this feature can be used to achieve the claimed security properties, which is the issue here.  What is the point of this feature if it cannot be used to achieve the claimed security properties?

Well, technically, you only claim "integrity protection", and that is achieved.  However, that seems to be because you're misusing the term to mean "authenticity protection".  If you only needed "integrity protection" there would be no need for signatures.

BTW, please do not mark conversations as resolved until they are resolved.

Reply:
I see that the new version you just force-pushed out adds setting `/proc/sys/fs/verity/require_signatures` to the test.  So I think you're acknowledging that there might be a problem here.  The issues that still remain, though, are:

* `systemd` doesn't actually enforce that fsverity is enabled on the files before using them.  So an attacker can replace the files with ones that don't have fsverity enabled, without detection.  It seems your position is that the system administrator needs to manually run `fsverity measure` on the files periodically.  That does not seem like a practical or secure approach.  The user of these files (`systemd`) needs to enforce that fsverity is enabled when reading from these files, if it's configured to do so.
* `require_signatures` isn't mentioned at all in the documentation for this feature.  This feature's documentation just has a vague mention that enabling fsverity with builtin signatures gives "signed fsverity integrity protection".  What that means is unclear.  It sounds a lot like authenticity protection, which if correct is actively misleading as there are additional steps that users would have to take to achieve this that the documentation does not mention at all.

Reply:
There's also the issue of FS_IOC_ENABLE_VERITY being used both with and without signatures.  It's not made clear under what circumstances `require_signatures` can even be used.

Reply:
Another oddity is that this feature's documentation name-drops IMA:

> fsverity will be used automatically on such files, to provide integrity protection and measurement in the IMA registry

But then for the signatures, it does **not** use IMA signatures, even though that is an option.  (IMA supports fsverity since Linux v5.19.)

So it's very unclear what anyone is supposed to do with this feature!

Reply:
I've already said elsewhere that I'll also add xattr support in parallel as soon as I have time, in this same PR.

> So it's very unclear what anyone is supposed to do with this feature!

It's quite simple really: use it if they want to for whatever reason, else ignore it, as with every other optional feature we provide. I understand this doesn't fit within google's narrow use case (implemented by proprietary, closed source software), but I am not looking for google's stamp of approval. google's views have been made abundantly clear, but this is not the LKML, so that's quite enough for now. Thank you for your contributions!